pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TopicTransactionBufferTest.java

```diff
@@ -405,4 +413,173 @@ private void assertGetLastMessageId(Consumer<?> consumer, MessageIdImpl expected
+        // Aborting the transaction and unload the topic and then redelivering unacked messages,
+        // only normal messages can be received.
+        transaction.abort().get(5, TimeUnit.SECONDS);
+        producer.newMessage().value(normalMsg).send();
+        admin.topics().unload(topic);
+        PersistentTopic persistentTopic2 = (PersistentTopic) pulsarServiceList.get(0).getBrokerService()
+                .getTopic(topic, false)
+                .get()
+                .get();
+        TopicTransactionBuffer topicTransactionBuffer2 = (TopicTransactionBuffer) persistentTopic2
+                .getTransactionBuffer();
+        Awaitility.await().untilAsserted(() -> {
+            Assert.assertEquals(topicTransactionBuffer2.getState(), TopicTransactionBufferState.State.Ready);
+        });
+        consumer.redeliverUnacknowledgedMessages();
+        for (int i = 0; i < 2; i++) {
+            Message<String> message = consumer.receive(5, TimeUnit.SECONDS);
+            Assert.assertEquals(message.getValue(), normalMsg);
+        }
+        Message<String> message = consumer.receive(5, TimeUnit.SECONDS);
+        Assert.assertNull(message);
+    }
+
+    /**
+     * Send some messages before transaction buffer ready and then send some messages after transaction buffer ready,
+     * these messages should be received in order.
+     */
+    @Test
+    public void testMessagePublishInOrder() throws Exception {
+        // 1. Prepare test environment.
+        this.pulsarServiceList.forEach(pulsarService ->  {
+            pulsarService.setTransactionExecutorProvider(new TransactionBufferTestProvider());
+        });
+        String topic = "persistent://" + NAMESPACE1 + "/testMessagePublishInOrder" + RandomUtils.nextLong();
+        admin.topics().createNonPartitionedTopic(topic);
+        PersistentTopic persistentTopic = (PersistentTopic) pulsarServiceList.get(0).getBrokerService()
+                .getTopic(topic, false)
+                .get()
+                .get();
+        @Cleanup
+        Producer<Integer> producer = pulsarClient.newProducer(Schema.INT32)
+                .topic(topic)
+                .create();
+        @Cleanup
+        Consumer<Integer> consumer = pulsarClient.newConsumer(Schema.INT32)
+                .topic(topic)
+                .subscriptionName("sub")
+                .subscribe();
+        Transaction transaction = pulsarClient.newTransaction()
+                .withTransactionTimeout(5, TimeUnit.HOURS)
+                .build().get();
+
+        // 2. Set a new future in transaction buffer as `transactionBufferFuture` to simulate whether the
+        // transaction buffer recover completely.
+        TransactionBufferTestImpl topicTransactionBuffer = (TransactionBufferTestImpl) persistentTopic
+                .getTransactionBuffer();
+        CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+        topicTransactionBuffer.setTransactionBufferFuture(completableFuture);
```

Comment:
Even though you set a non-completed future here, the future `transactionBuffer.publishFuture` is done when TransactionBuffer is created.